### DISCUSSION TITLE
What is the best practice for changing a child state in NavigationStack?

### DISCUSSION THREAD

**Discussion by MorningStarJ** [pointfreeco/swift-composable-architecture#3208] [Q&A] [answered]
First of all, thank you for the amazing work.

If I am using NavigationStack, A -> B -> C, and A is the reducer holding the NavigationStack path, how can I change the state of B in A when C changes some states and B needs to respond?
I don’t know if I’m missing something. I’ve been searching through Examples and Discussions for a long time and haven’t found a similar case.

```Swift
@Reducer
struct ÅFeature {
 struct State {
  var path: [Path] = []
  }
  ...
  enum Path {
    case a(AFeature)
    case b(BFeature)
    case c(CFeature)
  }

  enum Action {
    ...
    case path(StackActionOf<Path>)
  }
   
  var body: some ReducerOf<Self> {
    Reduce<State, Action> { state, action in

      switch action {
      case .path(.element(id: let cPathId, action: .c(.didChange(let cState)))):
        // What should I do?
       ...
        return .none

      case ...
      }
  }
}
```
First, I tried directly modifying the properties in bState, but if I understand correctly, due to the copy-on-write (COW) nature, I cannot directly modify the properties of B’s state in the navigationStack and make it effective because I only obtained a copy of bState.
```Swift
switch action {
      case .path(.element(id: let cPathId, action: .c(.didChange(let cState)))):
        // 1
        guard let bStackId = state.path.ids.first(where: { stackElementID in
             state.path[id: stackElementID]?.b != nil
           }),
           let bState = state.path[id: bStackId]?.b else {
          debugLog("can not find personalityCenterStackId")
          return .none
        }
        bState.property = ....
        return .none
```

I carefully read through this [discussion](https://github.com/pointfreeco/swift-composable-architecture/discussions/1952#discussioncomment-5167956), and based on mbrandonw’s response, I tried mapping the path action in A to return an action from B’s reducer,  I don’t know if I’m missing something, but it seems doesn't work.
```Swift
@Reducer
struct BFeature {
   struct State {
        ...
        mutating func update(…) -> Effect<BFeature.Action> {
        ...
        }
   }
}

switch action {
      case .path(.element(id: let cPathId, action: .c(.didChange(let cState)))):
        // 1
        guard let bStackId = state.path.ids.first(where: { stackElementID in
             state.path[id: stackElementID]?.b != nil
           }),
           let bState = state.path[id: bStackId]?.b else {
          debugLog("can not find personalityCenterStackId")
          return .none
        }
        return bState
        .update(...)
        .map { Action.path(.element(id: bStackId, action: .b($0)))}
```

Currently, I discovered swift-case-paths, which seems to have methods for directly modifying enum associated values, and it works.
```Swift
@Reducer
struct BFeature {
   struct State {
        ...
        mutating func update(…)  {
        ...
        }
   }
}

switch action {
  case .path(.element(id: let cPathId, action: .c(.didChange(let cState)))):
      guard let bStackId = state.path.ids.first(where: { stackElementID in
                   state.path[id: stackElementID]?.b != nil
                 }) else {
                debugLog("can not find personalityCenterStackId")
                return .none
              }
              state.path[id: bStackId]?.modify(\.b, yield: { bState in
                bState.update(cState: cState)
              })
      return .none
```
But it feels a bit strange, and I’m not sure if this is the best practice.
If anyone has any suggestions, I would greatly appreciate it.​

**Comment by mbrandonw** [comment, marked as answer]
Hi @MorningStarJ, your approaches are basically correct, though maybe a little verbose. Here is another approach:

```swift
guard let bIndex = state.path.firstIndex(where: { $0.is(\.b) })
else {
  return
}

state.path[bIndex].modify(\.b) {
  // Mutate $0
}
```

There `is(casePath)` function allows you to easily check if an enum matches a case, and it comes from our CasePaths library. And you can fetch the positional index rather than ID so that you can subscript directly into `path`. And then you can use `modify` with a case path to mutate the data inside the path.

**Reply by MorningStarJ** [reply]
Thank you very much, now I got it.

**Reply by MorningStarJ** [reply]
Hi, @mbrandonw, thanks for your solution, but it seems `state.path[bIndex].modify(\.b)` is causing a compilation error.
Cannot use mutating member on immutable value: subscript is get-only

swift-composable-architecture: v1.11.2
swift-case-path: v1.4.2
Xcode: Version 15.4 (15F31d)
Swift: 5.10

**Reply by mbrandonw** [reply]
Hi @MorningStarJ, sorry, my code was slightly wrong. Should be more like this:

```swift
guard let bIndex = state.path.firstIndex(where: { $0.is(\.b) })
else {
  return
}

let bID = state.path.ids[bIndex]
state.path[id: bID, case: \.b] = …
```

We should probably make some better APIs for juggling these IDs and indices in `StackState`.

**Reply by MorningStarJ** [reply]
Hi @mbrandonw, thank you for your response.I agree with your idea. If there is a simpler way to get a specific Path case ID from state.path, it might be more friendly.

**Reply by Young-Li-wxx** [reply]
> Hi @MorningStarJ, sorry, my code was slightly wrong. Should be more like this:
> 
> ```swift
> guard let bIndex = state.path.firstIndex(where: { $0.is(\.b) })
> else {
>   return
> }
> 
> let bID = state.path.ids[bIndex]
> state.path[id: bID, case: \.b] = …
> ```
> 
> We should probably make some better APIs for juggling these IDs and indices in `StackState`.

this can modify \.b's states,  but how can trigger \.b‘s action ?
